psh/wget/wget.c

```diff
@@ -0,0 +1,500 @@
+
+		tmp = memchr(common.buf, '\n', common.len);
+		/* We have not received a whole line yet, carry on reading */
+		if (tmp == NULL) {
+			common.readptr = common.buf + bytes;
+			continue;
+		}
+
+		/* Omit \r\n */
+		common.readptr = tmp + 1;
+
+		*(tmp - 1) = '\0';
+		tmp = strstr(common.buf, "HTTP");
+		if (tmp == NULL) {
+			return NULL;
+		}
+
+		tmp = strchr(tmp, ' ');
+		if (tmp == NULL) {
+			return NULL;
+		}
+
+		status = tmp + 1;
+	}
+
+
+	return status;
+}
+
+
+static void wget_cleanup(void)
+{
+	/* Cursor enable */
+	printf("\033[?25h");
+
+	fclose(common.outfile);
+	close(common.socket);
+	free(common.host);
+}
+
+
+int psh_wget(int argc, char **argv)
+{
+	const char *url;
+	const char *output = NULL;
+	char *status, *endptr;
+	size_t len;
+	int statnum;
+	int c;
+	struct timespec ts;
+	time_t begin, end, delta;
+
+	while ((c = getopt(argc, argv, "O:d")) != -1) {
+		switch (c) {
+			case 'O':
+				output = optarg;
+				break;
+			case 'd':
+				common.debug = 1;
+				break;
+			default:
+			case 'h':
+				wget_help();
+				return 0;
+		}
+	}
+
+	if (argc - optind != 1) {
+		fprintf(stderr, "URL missing!\n");
+		return 2;
+	}
+
+	url = argv[optind];
+
+	if (wget_parseUrl(url) != 0) {
+		return 2;
+	}
+
+	/* By default save the file to the current directory with the name from url */
+	if (output == NULL) {
+		output = common.filename;
+	}
+
+	common.outfile = fopen(output, "w+");
+	if (common.outfile == NULL) {
+		fprintf(stderr, "Fail to open file %s\n", output);
+		return 2;
+	}
+
+	/* Disable cursor */
+	printf("\033[?25l");
+
+	common.socket = wget_connect(common.host);
+	if (common.socket < 0) {
+		/* Cursor enable */
+		printf("\033[?25h");
+		fprintf(stderr, "wget: Fail to connect to host!\n");
+		fclose(common.outfile);
+		return 1;
+	}
+
+	if (wget_request() < 0) {
+		fprintf(stderr, "wget: HTTP request sending failed\n");
+		wget_cleanup();
+		return 1;
+	}
+
+	printf("HTTP request sent, awaiting response... ");
+	fflush(stdout);
+
+	status = wget_response();
+	if (status == NULL) {
+		fprintf(stderr, "wget: Failed to get response!\n");
+		wget_cleanup();
+		return 1;
+	}
+
+	statnum = strtol(status, &endptr, 10);
+	if (*endptr != ' ') {
+		fprintf(stderr, "wget: Wrong status format!\n");
+		wget_cleanup();
+		return 1;
+	}
+
+	printf("%s\n", status);
+
+	/* TODO: Handle other statuses properly */
+	if (statnum < 200 || statnum >= 300) {
+		wget_cleanup();
+		return 1;
+	}
+
+	len = wget_parsehdrs();
+	printf("Length: %d\n", len);
+	if (len == 0) {
+		printf("Nothing to be copied\n");
+		wget_cleanup();
+		return 0;
+	}
+
+	printf("Saving to: '%s'\n", output);
+
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	begin = ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
+	len = wget_file(len);
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	end = ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
+	delta = end - begin;
+
+	printf("Downloaded %d bytes in %lld.%03llds\n", len, delta / 1000000, (delta - (delta / 1000000) * 1000000) / 1000);
```

Comment:
`wget_file()` is type of `int` and on error return `-1`.
You are printing time as `ss.ms` so maybe `begin, end, delta` should contain ms.
```suggestion
	int ret;
	ret = wget_file(len);
	if (ret >= 0) {
		clock_gettime(CLOCK_MONOTONIC, &ts);
		end = ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
		delta = end - begin;

		printf("Downloaded %d bytes in %lld.%03llds\n", ret, delta / 1000000, (delta - (delta / 1000000) * 1000000) / 1000);
	}
```